### PR TITLE
Close workers gracefully

### PR DESCRIPTION
**Pull request by mrocklin**
This allows workers to optionally terminate themselves gracefully after a predetermined time.  This can be helpful in a few contexts:

1.  We receive a SIGINT, and know that we need to clean up quickly (though note that the signal handlers are not implemented as part of this PR (cc @jni))
2.  We know that we'll be kicked off at a certain time, such as in one hour from now, as is often specified by HPC job schedulers (cc @lesteve @guillaumeeb @jhamman )
3.  We just want to refresh our workers every once in a while, because we know that our code leaks some memory .  Fixes https://github.com/dask/distributed/issues/2861 

This is configurable as keywords to the `Worker` or `Nanny` classes, in config values, or with CLI.  Here is an example with CLI.

### Restart to clear state

```
dask-worker scheduler:8786 --lifetime 1hr --lifetime-restart --lifetime-stagger 5m
```

This will kill the worker roughly 1 hour from now +- a range of 5 minutes (to avoid killing all of our workers at the same time).  It will also allow that worker to be restarted afterwards

### Restart to avoid walltime death

```
dask-worker scheduler:8786 --lifetime 58m 
```

Here we don't try to restart the worker (no point) and we choose a time a bit before our 60m walltime.

Also cc @jacobtomlinson @jcrist for general deployment information

For folks who want to review.  I recommend going commit-by-commit (it should be fairly clean).  (this is also a decently simple/educational PR to review)

### CODE REVIEW

**Comment by mrocklin** [comment]
Merging this tomorrow if there are no further comments (Although comments would be quite welcome here)

**Comment by TomAugspurger** [comment]
> though note that the signal handlers are not implemented as part of this PR

Do you anticipate this making fixing https://github.com/dask/distributed/issues/2788 easier? https://github.com/dask/distributed/pull/2892 makes it sound like "yes".

**Comment by mrocklin** [comment]
> Do you anticipate this making fixing #2788 easier?

It's the action that I think we would want to take when handling a SIGINT.

**Comment by mrocklin** [comment]
Merging this shortly if there are no further comments

File: distributed/worker.py [hunk 1]
```diff
@@ -252,6 +252,16 @@ class Worker(ServerNode):
         Resources that this worker has like ``{'GPU': 2}``
     nanny: str
         Address on which to contact nanny, if it exists
+    lifetime: str
+        Amount of time like "1 hour" after which we gracefully shut down the worker.
+        This defaults to None, meaning no explicit shutdown time.
+    lifetime_stagger: str
+        Amount of time like "5 minutes" to stagger the lifetime value
+        The actual lifetime will selected uniformly at random between
```
Comment (guillaumeeb) [inline distributed/worker.py]:
```suggestion
        The actual lifetime will be selected uniformly at random between
```

File: distributed/cli/dask_worker.py [hunk 2]
```diff
@@ -175,6 +175,30 @@
 @click.option(
     "--dashboard-prefix", type=str, default="", help="Prefix for the dashboard"
 )
+@click.option(
+    "--lifetime",
+    type=str,
+    default="",
+    help="If provided, shut down the worker after this duration. "
+    "Note that if combined with --nanny (default) "
+    "the worker will restart after this time",
```
Comment (guillaumeeb) [inline distributed/cli/dask_worker.py]: Even with lifetime_restart to false as default?

Comment (mrocklin) [reply distributed/cli/dask_worker.py]: Thanks.  I've removed this.  I think that it's now clear with the `--lifetime-restart` keyword just below.

**Comment by lesteve** [comment]
Just a quick comment (I don't think this is a problem for this PR, since the signal handling is only part of it AFAICT): for SGE the signal you get when the job exceed the walltime is a `SIGKILL` (e.g. `kill -9`). Not sure for other schedulers.

**Comment by jacobtomlinson** [comment]
Ouch! AFAIK Slurm and PBS send a `SIGTERM` shortly before the `SIGKILL` to give the process a chance to exit cleanly.

**Comment by lesteve** [comment]
Good to know that other schedulers are a bit nicer regarding job termination.

I found [this](http://disbauxes.upc.es/code/analysis/automatic-job-resubmit-using-sun-grid-engine/) which seems to indicate that for SGE `SIGKILL` is used by default but that you can configure it to be something else.

**Pull request merged**